### PR TITLE
CSApply: Fix several issues with non-failable to failable/throwing initializer delegation/chaining diagnostics

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
* Delegated-to `Optional` ctors were always handled as if they were failable, resulting in false-positive delegation errors.
* Delegations via `try?` were not diagnosed if the called ctor had IUO failability on top of being `throws`. SILGen would then handle the `try?` as if it was a `try!`.
* Delegations via `try?` were diagnosed with the wrong message if the `try?` was nested inside a `try!`, e.g. `try! try self.init(nonFailable:)`
* If there are issues with both `try?` and failability of the called initializer, diagnose both.

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -3636,57 +3636,89 @@ namespace {
...
-        if (isError && SuppressDiagnostics)
-          return nullptr;
+        // Look through 'try', 'try!', and identity expressions.
+        subExpr = subExpr->getValueProvidingExpr();
 
-        auto &ctx = cs.getASTContext();
+        // Diagnose if we find a 'try?'.
+        // FIXME: We could put up with occurences of 'try?' if they do not apply
+        // directly to the called ctor, e.g. 'try? try self.init()', or if the
+        // called ctor isn't throwing.
```
Comment (AnthonyLatsis) [inline lib/Sema/CSApply.cpp:3661]: What do you think, is this worth addressing later?

**Review by xedin — APPROVED** [review]
Thank you!

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Comment by AnthonyLatsis** [comment]
@swift-ci please test source compatibility

**Comment by AnthonyLatsis** [comment]
Those KingFisher failures also appear in earlier builds.

**Comment by AnthonyLatsis** [comment]
@xedin Thanks!

**Pull request merged**